sql/expression/in.go

```diff
@@ -280,47 +280,66 @@ func hashOfLiteral(l *Literal, t sql.Type) (uint64, error) {
 // hashOfTuple will recursively hash a Tuple tree with Literal leaves
 func hashOfTuple(tup Tuple, t sql.TupleType) (uint64, error) {
 	hash := xxhash.New()
+	var x interface{}
+	var err error
 	for i, el := range tup {
 		switch v := el.(type) {
 		case *Literal:
-			converted, err := t[i].Promote().Convert(v.value)
-			if err != nil {
-				return 0, err
-			}
-			if _, err := hash.Write([]byte(fmt.Sprintf("%#v,", converted))); err != nil {
-				return 0, err
-			}
+			x = v.value
 		default:
 			return 0, ErrCantHashNestedExpression.New(v)
 		}
+		x, err = t[i].Promote().Convert(x)
+		if err != nil {
+			return 0, err
+		}
+		if _, err := hash.Write([]byte(fmt.Sprintf("%#v,", x))); err != nil {
+			return 0, err
+		}
 	}
 	return hash.Sum64(), nil
 }
 
-func normalizeLeft(ctx *sql.Context, expr sql.Expression, row sql.Row) (sql.Expression, error) {
-	switch e := expr.(type) {
-	case Tuple:
-		return TransformUp(e, func(expr sql.Expression) (sql.Expression, error) {
-			switch e := expr.(type) {
-			case *GetField:
-				v, err := e.Eval(ctx, row)
-				if err != nil {
-					return nil, err
-				}
-				return NewLiteral(v, e.Type()), nil
-			default:
-				return e, nil
+func evalLeftExpr(ctx *sql.Context, expr sql.Expression, row sql.Row) (sql.Expression, error) {
+	return TransformUp(expr, func(expr sql.Expression) (sql.Expression, error) {
+		switch e := expr.(type) {
+		case Tuple:
+			return e, nil
+		case *Literal:
+			return e, nil
+		case *GetField, *IsTrue, *IsNull, *Convert, Comparer, sql.FunctionExpression, *Arithmetic:
+			v, err := e.Eval(ctx, row)
+			if err != nil {
+				return nil, err
 			}
-		})
-	case *Literal:
-		return e, nil
-	case *GetField:
-		v, err := e.Eval(ctx, row)
-		if err != nil {
-			return nil, err
+			return NewLiteral(v, e.Type()), nil
+		default:
+			return nil, ErrHashInLeftNotSupported.New(e)
```

Comment:
Same comment here

Also you're taking a whitelist approach here when it should ideally be a blacklist. This is going to break as we add new expression types. There needs to be an iron-clad contract between the analyzer and the execution logic, and what you have here now makes me nervous. This should be as simple as calling `left.Eval()` and (not sure why this is necessary) wrapping it in a literal.

Reply:
the existing code was just a bit easier because we have the framework to manipulate and hash trees of expressions. would i cast tuples to lists for hashing? ex: `Tuple(Literal(1), Literal(0)) -> [1,0]` -> ` hash([1,0])`?

Reply:
rewrote to simplify hashIn logic, assuming whitelisted arguments and without recursive eval/hasher